### PR TITLE
verify jwt token

### PR DESCRIPTION
**Pull request by aviled**

### CODE REVIEW

File: src/functionApp/SmartDetectorsFunctionApp/Analyze.cs [hunk 1]
```diff
@@ -42,7 +43,8 @@ static Analyze()
             ThreadPool.SetMinThreads(100, 100);
 
             Container = DependenciesInjector.GetContainer()
-                    .InjectAnalysisDependencies(withChildProcessRunner: true);
+                .InjectAnalysisDependencies(withChildProcessRunner: true)
+                .RegisterType<IAuthorizationManagementClient, AuthorizationManagementClient>();
         }
```
Comment (YairoR) [inline src/functionApp/SmartDetectorsFunctionApp/Analyze.cs]: can you fix the indentation? the dot should be under the . of GetContainer()

File: src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs [hunk 2]
```diff
@@ -6,44 +6,35 @@
...
     /// <summary>
     /// This class is responsible to manage authorization for the Smart Detectors Monitoring Appliance
     /// </summary>
     public class AuthorizationManagementClient : IAuthorizationManagementClient
     {
-        /// <summary>
-        /// The dependency name, for telemetry
-        /// </summary>
-        private const string DependencyName = "RBAC";
+        private const string MicrosoftTenant = "72f988bf-86f1-41af-91ab-2d7cd011db47";
```
Comment (YairoR) [inline src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs]: >MicrosoftTenant [](start = 29, length = 15)

You don't want to validate the target resource also? (00000002-0000-0000-c000-000000000000)

File: src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs [hunk 3]
```diff
@@ -61,6 +63,14 @@ static SmartDetector()
 
                 try
                 {
+                    // Check authorization
+                    var authorizationManagementClient = childContainer.Resolve<IAuthorizationManagementClient>();
```
Comment (YairoR) [inline src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs]: >authorizationManagementClient [](start = 24, length = 29)

The logic is correct, but the way we are using it doesn't look nice.
there are 2 options here:
1. Create a base class which will run the method logic only if user is authorized. 
2. Create a Function-Filter which will do the validation ([https://github.com/Azure/azure-webjobs-sdk/wiki/Function-Filters](https://github.com/Azure/azure-webjobs-sdk/wiki/Function-Filters)).
Also this feature is not GA yet, I think we can go for it.

Comment (aviled) [reply src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs]: function filter sounds like an amazing idea!
will do on Sunday.
Love, 
Aviram

---
In reply to: [176530730](https://github.com/liadtal/Smart-Alerts/pull/114#discussion_r176530730) [](ancestors = 176530730)

File: src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs (hunk 2 above)
Comment (aviled) [reply src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs]: nope, talked offline

---
In reply to: [176530220](https://github.com/liadtal/Smart-Alerts/pull/114#discussion_r176530220) [](ancestors = 176530220)

**Comment by aviled** [comment]
>             using (IUnityContainer childContainer = Container.CreateChildContainer().WithTracer(log, true))

removed

---
Refers to: src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs:59 in efd52a0. [](commit_id = efd52a098250808141558cf6a37c909c89ecc09d, deletion_comment = False)

File: src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs (hunk 3 above)
Comment (liadtal) [reply src/functionApp/SmartDetectorsFunctionApp/SmartDetector.cs]: Another option (if the awesome Filters feature doesn't work) is to move this logic to shared

---
In reply to: [176535902](https://github.com/liadtal/Smart-Alerts/pull/114#discussion_r176535902) [](ancestors = 176535902,176530730)

File: src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs [hunk 4]
```diff
@@ -61,43 +52,42 @@ public AuthorizationManagementClient(ITracer tracer, IHttpClientWrapper httpClie
...
-            PermissionGetResult permissionGetResult = JsonConvert.DeserializeObject<PermissionGetResult>(permissionsString);
+            string appId = claimsPrincipal.Claims.FirstOrDefault(claim => claim.Type == "appid")?.Value;
 
-            if (permissionGetResult.Permissions == null || !permissionGetResult.Permissions.Any() || !permissionGetResult.Permissions.Any(p => p.Actions.Any()))
+            if (appId == SmartDetectorRuleEngineClientAppId)
             {
-                this.tracer.TraceWarning("No permissions were found for the token");
-                return false;
+                return true;
             }
```
Comment (liadtal) [inline src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs]: You have 2 string comparisons here - should you compare case-insensitive?

File: src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs [hunk 5]
```diff
@@ -6,44 +6,35 @@
...
     /// This class is responsible to manage authorization for the Smart Detectors Monitoring Appliance
     /// </summary>
     public class AuthorizationManagementClient : IAuthorizationManagementClient
     {
-        /// <summary>
-        /// The dependency name, for telemetry
-        /// </summary>
-        private const string DependencyName = "RBAC";
+        private const string MicrosoftTenant = "72f988bf-86f1-41af-91ab-2d7cd011db47";
+        private const string SmartDetectorRuleEngineClientAppId = "62c7f910-f075-4421-b6a0-47dee29bf7ef";
```
Comment (liadtal) [inline src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs]: Move to application settings

File: src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs (hunk 4 above)
Comment (aviled) [reply src/functionApp/SmartDetectorsFunctionApp/Authorization/AuthorizationManagementClient.cs]: removed this class

---
In reply to: [176933727](https://github.com/liadtal/Smart-Alerts/pull/114#discussion_r176933727) [](ancestors = 176933727)

File: src/deployment/MonitoringApplianceDeployment/mainTemplate.json [hunk 6]
```diff
@@ -189,6 +190,31 @@
...
+            "enabled": true,
+            "unauthenticatedClientAction": "RedirectToLoginPage",
+            "tokenStoreEnabled": true,
+            "defaultProvider": "AzureActiveDirectory",
+            "clientId": "[variables('ruleEngineClientId')]",
+            "issuer": "https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47",
+            "allowedAudiences": [
+              "00000002-0000-0000-c000-000000000000"
+            ]
+          }
```
Comment (liadtal) [inline src/deployment/MonitoringApplianceDeployment/mainTemplate.json]: COOL !

**Pull request merged**